### PR TITLE
[Merged by Bors] - ET-3987 change-persona-tests

### PR DESCRIPTION
**Pull request by krajewskiML**
### Reference

[ET-3987](https://xainag.atlassian.net/browse/ET-3987?atlOrigin=eyJpIjoiMzE4YWZlNTQ5ZGQ1NDkyMjkwNzBjNGJjZWU0MWI1NWQiLCJwIjoiaiJ9)

### Summary
- adds saturation test which measures how does CoI exploit topics
- adds persona test with random sampling from hot news
- adds random seeds to tests
- allows different calculation of NDCG

[ET-3987]: https://xainag.atlassian.net/browse/ET-3987?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by PanicButtonPressed** [comment]
Let's rebase and resolve conflicts

File: web-api/src/mind.rs [hunk 1]
```diff
@@ -552,6 +676,326 @@ async fn run_user_benchmark() -> Result<(), Error> {
...
+                let scores = score_documents(
+                    &documents,
+                    interests,
+                    grid_search_config.semi_interesting_flag,
+                );
+                let ndcgs_iteration = ndcg(&scores, &grid_search_config.nranks);
+                //save scores to results array
+                for (i, ndcg) in ndcgs_iteration.iter().enumerate() {
+                    ndcgs_all_configs[[config_idx, idx, iter, i]] = *ndcg;
+                }
```
Comment (janpetschexain) [inline web-api/src/mind.rs:976]: 🔧 similarly
```suggestion
                ndcgs_all_configs
                    .slice_mut(ndarray::s![config_idx, idx, iter, :])
                    .assign(&Array::from(ndcgs_iteration));
```

File: web-api/src/mind.rs [hunk 2]
```diff
@@ -552,6 +676,326 @@ async fn run_user_benchmark() -> Result<(), Error> {
...
+                    .interact(
+                        user_id,
+                        &personalised_documents
+                            .iter()
+                            .zip(scores.iter())
+                            .filter(|(_, &score)| {
+                                (score - 2.0).abs() < 0.001
+                                    && rng.gen_bool(grid_search_config.click_probability)
+                            })
+                            .map(|(id, _)| (id.clone(), state.time - Duration::days(0)))
```
Comment (janpetschexain) [inline web-api/src/mind.rs:988]: 🔧 similarly
```suggestion
                            .zip(scores)
                            .filter_map(|(id, score)| {
                                ((score - 2.0).abs() < 0.001
                                    && rng.gen_bool(grid_search_config.click_probability))
                                .then(|| (id.clone(), state.time))
                            })
```

**Review by janpetschexain — COMMENTED** [review]
i mainly left rust related comments, the test logic looks good

:art: there is quite a lot of code repetition in the various tests now, maybe in a follow up pr we can modularize the repeated code and reuse it for better maintainability

File: web-api/src/mind.rs [hunk 3]
```diff
@@ -296,6 +389,22 @@ impl DocumentProvider {
...
 }
+#[derive(Debug, derive_more::Deref, Deserialize)]
+struct SpecificTopics(Vec<String>);
+
+impl SpecificTopics {
+    fn new(path: &str) -> Result<Self, Error> {
+        let file = File::open(path)?;
+        let json = serde_json::from_reader::<_, serde_json::Value>(file)?;
+        let list = json.as_array().unwrap();
+        // iterate over vector and create a pair of strings
```
[the remark this answers is not in the document]
Comment (krajewskiML) [reply web-api/src/mind.rs:400]: You are right, I have started to implement the solution using two strings but then I realised that I do not need them to be split and I left the comment, I will remove it

File: web-api/src/mind.rs [hunk 4]
```diff
@@ -552,6 +676,326 @@ async fn run_user_benchmark() -> Result<(), Error> {
...
 
+/// The function panics if the provided filenames are not correct.
+#[tokio::test]
+#[ignore]
+async fn run_saturation_test() -> Result<(), Error> {
+    // load list of possible specific topics from file (need to create it)
+    let specific_topics = SpecificTopics::new(
+        "/Users/maciejkrajewski/CLionProjects/xayn_discovery_engine/web-api/src/bin/topics.json",
+    )?;
+    let document_provider = DocumentProvider::new("/Users/maciejkrajewski/CLionProjects/xayn_discovery_engine/web-api/src/bin/news_cleaned.tsv")?;
```
[the remark this answers is not in the document]
Comment (krajewskiML) [reply web-api/src/mind.rs:687]: Yeah, absolutely I forgot about that

File: web-api/src/mind.rs [hunk 5]
```diff
@@ -552,6 +676,326 @@ async fn run_user_benchmark() -> Result<(), Error> {
...
+        let documents = document_provider.get_all_interest(&[full_category.clone()]);
+        let user_id = UserId::new(full_category).unwrap();
+        // get random document from the topic
+        let document = documents.choose(&mut rng).unwrap();
+
+        // interact with the document
+        state
+            .interact(
+                &user_id,
+                &[(document.id.clone(), state.time - Duration::days(0))],
```
[the remark this answers is not in the document]
Comment (krajewskiML) [reply web-api/src/mind.rs:718]: Yeah I know, but I guess this value will be changed according to the test that we will be run so I left it as a reminder

File: web-api/src/mind.rs [hunk 6]
```diff
@@ -552,6 +676,326 @@ async fn run_user_benchmark() -> Result<(), Error> {
     Ok(())
 }
 
+/// The function panics if the provided filenames are not correct.
+#[tokio::test]
+#[ignore]
+async fn run_saturation_test() -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (krajewskiML) [reply web-api/src/mind.rs:682]: Good point, I will rename it

File: web-api/src/mind.rs [hunk 7]
```diff
@@ -225,6 +277,43 @@ impl Default for StateConfig {
     }
 }
 
+// The results of iteration of the saturation benchmark
```
Comment (janpetschexain) [inline web-api/src/mind.rs:280]: :art: please note the difference between a double `//` for comments and a triple `///` for doc comments, only the tripple gets picked up for the doc generation. similarly for the other type doc comments below.
```suggestion
/// The results of iteration of the saturation benchmark
```

**Review by janpetschexain — APPROVED** [review]
lgtm

File: web-api/src/mind.rs (hunk 7 above)
Comment (krajewskiML) [reply web-api/src/mind.rs:280]: Ah yes, I will change it, I must have skipped this one by accident

**Comment by krajewskiML** [comment]
bors merge

**Pull request closed without merging**